cms/src/main/java/uk/nhs/digital/externalstorage/modules/S3ConnectorServiceRegistrationModuleParams.java

```diff
@@ -1,74 +1,91 @@
-    private final String systemPropertyKey;
-    private final String jcrModulePropertyKey;
-    private final Function<String, T> fromStringConverter;
-    private final T defaultValue;
-
-    private S3ConnectorServiceRegistrationModuleParams(final String systemPropertyKey,
-                                                       final String jcrModulePropertyKey,
-                                                       final Function<String, T> fromStringConverter,
-                                                       final T defaultValue
-    ) {
-        this.systemPropertyKey = systemPropertyKey;
-        this.jcrModulePropertyKey = jcrModulePropertyKey;
-        this.fromStringConverter = fromStringConverter;
-        this.defaultValue = defaultValue;
+import javax.jcr.Node;
+import javax.jcr.RepositoryException;
+
+public class S3ConnectorServiceRegistrationModuleParams {
+    private String s3Bucket;
+    private String s3Region;
+    private String s3Endpoint;
+
+    private long downloadsMaxConcurrentCount;
+    private long downloadsShutdownTimeoutInSecs;
+
+    private long uploadsMaxConcurrentCount;
+    private long uploadsShutdownTimeoutInSecs;
+
+    // SYSTEM PROPERTY KEYS
+    public static final String KEY_AWS_BUCKET =  "externalstorage.aws.bucket";
+    public static final String KEY_AWS_REGION =  "externalstorage.aws.region";
+    public static final String KEY_AWS_S3_ENDPOINT =  "externalstorage.aws.s3.endpoint";
+
+    // JCR PROPERTIES KEYS
+    public static final String KEY_DOWNLOAD_MAX_CONC_COUNT =  "maxConcurrentDownloadsCount";
+    public static final String KEY_DOWNLOAD_SHUTDOWN_TIMEOUT_IN_SECS =  "downloadShutdownTimeoutInSecs";
+    public static final String KEY_UPLOAD_MAX_CONC_COUNT =  "maxConcurrentUploadsCount";
+    public static final String KEY_UPLOAD_SHUTDOWN_TIMEOUT_IN_SECS =  "uploadShutdownTimeoutInSecs";
+
+    S3ConnectorServiceRegistrationModuleParams(final Node moduleConfigNode) {
+
+        // System properties set via command line
+        s3Bucket = System.getProperty(KEY_AWS_BUCKET, "");
+        s3Region = System.getProperty(KEY_AWS_REGION, "");
+        s3Endpoint = System.getProperty(KEY_AWS_S3_ENDPOINT, "");
+
+        // JCR properties set by YAML config
+        // (hippo:configuration/hippo:modules/s3ConnectorRegistrationModule/hippo:moduleconfig)
+        downloadsMaxConcurrentCount = getLongJcrPropertyValue(moduleConfigNode, KEY_DOWNLOAD_MAX_CONC_COUNT);
+        downloadsShutdownTimeoutInSecs = getLongJcrPropertyValue(moduleConfigNode, KEY_DOWNLOAD_SHUTDOWN_TIMEOUT_IN_SECS);
+        uploadsMaxConcurrentCount = getLongJcrPropertyValue(moduleConfigNode, KEY_UPLOAD_MAX_CONC_COUNT);
+        uploadsShutdownTimeoutInSecs = getLongJcrPropertyValue(moduleConfigNode, KEY_UPLOAD_SHUTDOWN_TIMEOUT_IN_SECS);
+    }
+
+    private long getLongJcrPropertyValue(Node moduleConfigNode, String jcrPropertyKey) {
+        try {
+            return moduleConfigNode.getProperty(jcrPropertyKey).getLong();
```

Comment:
Hmm, it seems that my suggestion to use `getLong()` here and to set the type of the timeout and count fields as `long` resulted in a lot of casting to `int` and back to `long` outside of this class. Sorry about that.

How about:
* changing the type of those fields and of values returned by the corresponding getter methods to `int`,
* eliminating all the casting outside of this class,
* if needed, only doing the casting once, within this mehtod, changing it to return `int` and renaming it to `getIntJcrPropertyValue`.

It seems that it should also be possible to avoid casting even within this method by using (needs testing to find out whether `getDecimal()` will work correctly for these properties):

```java
return moduleConfigNode.getProperty(jcrPropertyKey).getDecimal().intValue();
```